File: utils/lcd.py
from spi import *


ILI9341_PWCTR1 = 0xC0
ILI9341_PWCTR2 = 0xC1
ILI9341_VMCTR1 = 0xC5
ILI9341_VMCTR2 = 0xC7
ILI9341_MADCTL = 0x36
ILI9341_PIXFMT = 0x3A
ILI9341_FRMCTR1 = 0xB1
ILI9341_DFUNCTR = 0xB6
ILI9341_GAMMASET = 0x26
ILI9341_GMCTRP1 = 0xE0
ILI9341_GMCTRN1 = 0xE1
ILI9488_CASET =  0x2A
ILI9488_PASET =  0x2B
ILI9488_RAMWR =  0x2C

init_commands = [[0xEF, 0x03, 0x80, 0x02], 
[0xCF, 0x00, 0XC1, 0X30],
[0xED, 0x64, 0x03, 0X12, 0X81],
[0xE8, 0x85, 0x00, 0x78],
[0xCB, 0x39, 0x2C, 0x00, 0x34, 0x02],
[0xF7, 0x20],
[0xEA, 0x00, 0x00],
[ILI9341_PWCTR1, 0x23], #Power control
[ILI9341_PWCTR2, 0x10], # Power control
[ILI9341_VMCTR1, 0x3e, 0x28], # VCM control
[ILI9341_VMCTR2, 0x86], # VCM control2
[ILI9341_MADCTL, 0x48], # Memory Access Control
[ILI9341_PIXFMT, 0x55],
[ILI9341_FRMCTR1, 0x00, 0x18],
[ILI9341_DFUNCTR, 0x08, 0x82, 0x27], #Display Function Control
[0xF2, 0x00], #Gamma Function Disable
[ILI9341_GAMMASET, 0x01], #Gamma curve selected
[ILI9341_GMCTRP1, 0x0F, 0x31, 0x2B, 0x0C, 0x0E, 0x08,
0x4E, 0xF1, 0x37, 0x07, 0x10, 0x03, 0x0E, 0x09, 0x00], #Set Gamma
[ILI9341_GMCTRN1, 0x00, 0x0E, 0x14, 0x03, 0x11, 0x07,
	0x31, 0xC1, 0x48, 0x08, 0x0F, 0x0C, 0x31, 0x36, 0x0F, # Set Gamma
3, 0xb1, 0x00, 0x10, # FrameRate Control 119Hz
0]]


def init(init_commands): 
    for i in init_commands:
        control_dcx(State.OFF)
        send_cmd(i[0])
        control_dcx(State.ON)
        spi_write_data(i[1:],0,len(i[1:]))



def setAddrWindow(x0, y0, x1, y1) :
    control_dcx(State.OFF)
    send_cmd(ILI9488_CASET);  #Column addr set
    control_dcx(State.ON)
    addr_commands.append(x0 >> 8)
    addr_commands.append(x0 & 0xff)
    addr_commands.append(x1 >> 8)
    addr_commands.append(x1 & 0xff)
    spi_write_data(addr_commands, 0, len(addr_commands));     #XEND
    addr_commands.clear()

    control_dcx(State.OFF)
    send_cmd(ILI9488_PASET);  #Row addr set
    control_dcx(State.ON)
    addr_commands.append(y0 >> 8)
    addr_commands.append(y0 & 0xff)
    addr_commands.append(y1 >> 8)
    addr_commands.append(y1 & 0xff)
    spi_write_data(addr_commands, 0, len(addr_commands))

    send_cmd(ILI9488_RAMWR); #write to RAM


def write16BitColor(color) :
    r = (color & 0xF800) >> 11;
    g = (color & 0x07E0) >> 5;
    b = color & 0x001F;

    color = []

    r = (r * 255) // 31;
    g = (g * 255) // 63;
    b = (b * 255) // 31;

    color.append(r)
    color.append(g)
    color.append(b)
    control_dcx(State.OFF)
    spi_write_data(color,0 , len(color));
    control_dcx(State.ON)

def drawPixel(x, y, color) :

    setAddrWindow(x,y,x+1,y+1);

    control_dcx(State.ON)

    write16BitColor(color);

    control_dcx(State.OFF)



p.lcd()

File: src/lcd.c
#include "lcd.h"
#include "spi.h"
#include "types.h"
#include "uart.h"
#include "utils.h"
#include "gpio.h"

s16 _width = ILI9341_TFTWIDTH;
s16 _height = ILI9341_TFTHEIGHT;
s16 addr_row = (short) 0xFFFF;
s16 addr_col = (short) 0xFFFF;
s16 old_x1 = (short) 0xffff, old_x2 = (short) 0xffff;
s16 old_y1 = (short) 0xffff, old_y2 = (short) 0xffff;

u8 init_commands[] = {
	0xEF, 3, 0x03, 0x80, 0x02, 
	0xCF, 3, 0x00, 0xC1, 0x30,
	0xED, 4, 0x64, 0x03, 0x12, 0x81,
	0xE8, 3, 0x85, 0x00, 0x78,
	0xCB, 5, 0x39, 0x2C, 0x00, 0x34, 0x02,
	0xF7, 1, 0x20,
	0xEA, 2, 0x00, 0x00,
	ILI9341_PWCTR1, 1, 0x23, //#Power control
	ILI9341_PWCTR2, 1, 0x10, //# Power control
	ILI9341_VMCTR1, 2, 0x3e, 0x28, //# VCM control 
	ILI9341_VMCTR2, 1, 0x86, //# VCM control2
	ILI9341_MADCTL, 1, 0x48, //# Memory Access Control
	ILI9341_PIXFMT, 1, 0x55,
	ILI9341_FRMCTR1, 2, 0x00, 0x18, 
	ILI9341_DFUNCTR, 3,  0x08, 0x82, 0x27, //#Display Function Control
	0xF2, 1, 0x00, //#Gamma Function Disable
	ILI9341_GAMMASET, 1, 0x01, //#Gamma curve selected
	ILI9341_GMCTRP1, 15, 0x0F, 0x31, 0x2B, 0x0C, 0x0E, 0x08,
	0x4E, 0xF1, 0x37, 0x07, 0x10, 0x03, 0x0E, 0x09, 0x00, //#Set Gamma
	ILI9341_GMCTRN1, 15, 0x00, 0x0E, 0x14, 0x03, 0x11, 0x07, 
	0x31, 0xC1, 0x48, 0x08, 0x0F, 0x0C, 0x31, 0x36, 0x0F, //# Set Gamma
	0xb1, 2, 0x00, 0x10, //# FrameRate Control 119Hz
	ILI9341_SLPOUT, 0,// Exit Sleep
	ILI9341_DISPON, 0 // Display on
};

void init_lcd() {
	char numArgs;
	u8 comm;
	u8 *p = init_commands;
	for (u8 i = 0; i <= size_init; i++) {
		comm = *p;
		writeCommand(*(p++));
		if (comm == ILI9341_DISPON) break;
		numArgs = *(p++);
		while (numArgs-- > 0) {
			//spi_write_data((p+2), 0, numArgs);
			writedata(*(p++));
		} 
	
	}
	delay_us(500000);
}

void setAddrWindow(s16 x1, s16 y1, s16 x2, s16 y2) {
	if (x1 != old_x1 || x2 != old_x2) {
		writeCommand(ILI9341_CASET); // Column address set
		writedata(x1 >> 8);
		writedata(x1 & 0xff);
		writedata(x2 >> 8);
		writedata(x2 & 0xff);
		old_x1 = x1;
		old_x2 = x2;
	}
	if (y1 != old_y1 || y2 != old_y2) {
		writeCommand(ILI9341_PASET); // Row address set
		writedata(y1 >> 8);
		writedata(y1 & 0xff);
		writedata(y2 >> 8);
		writedata(y2 & 0xff);
		old_y1 = y1;
		old_y2 = y2;
	}
	writeCommand(ILI9341_RAMWR); // Write to RAM
}

void drawPixel(s16 x, s16 y, u16 color) {
	setAddrWindow(x, y, x+1, y+1);
	writedata(color >> 8);
	writedata(color & 0xff);
}

void fillScreen(u16 color) {
	fillRect(0, 0, _width, _height, color);
}

u16 color565(u8 r, u8 g, u8 b) {
	return ((r & 0xF8) << 8) | ((g & 0xFC) << 3) | (b >> 3);
}

void fillRect(s16 x, s16 y, s16 w, s16 h, u16 color) {
	if((x >= _width) || (y >= _height)) return;
	if((x + w - 1) >= _width)  w = _width  - x;
	if((y + h - 1) >= _height) h = _height - y;

	setAddrWindow(x, y, w, h);
	s32 total = 2*w*h;
	while (total-- > 0) { 
		writedata(color);
	}

}

void drawarr(u32 height, u32 width, u16 arr[height][width]) {
	for(int y=0; y < height; y++) {
		for(int x=0; x < width; x++ ) {
			drawPixel(x, y, arr[y][x]);
		}
	}
}

void setRotation(u8 m) {
	u8 rotation = m % 4; // can't be higher than 3
	writeCommand(ILI9341_MADCTL);
	switch (rotation) {
		case 0:
			m = (ILI9341_MADCTL_MX | ILI9341_MADCTL_BGR);
			_width = ILI9341_TFTWIDTH;
			_height = ILI9341_TFTHEIGHT;
			break;
		case 1:
			m = (ILI9341_MADCTL_MV | ILI9341_MADCTL_BGR);
			_width = ILI9341_TFTHEIGHT;
			_height = ILI9341_TFTWIDTH;
			break;
		
		case 2:
			m = (ILI9341_MADCTL_MY | ILI9341_MADCTL_BGR);
			_width = ILI9341_TFTWIDTH;
			_height = ILI9341_TFTHEIGHT;
			break;
		case 3:
			m = (ILI9341_MADCTL_MX | ILI9341_MADCTL_MY | ILI9341_MADCTL_MV | ILI9341_MADCTL_BGR);
			_width = ILI9341_TFTHEIGHT;
			_height = ILI9341_TFTWIDTH;
			break;
	}
	writedata(m);
	old_x1 = (short) 0xffff;
	old_x2 = (short) 0xffff;
	old_y1 = (short) 0xffff;
	old_y2 = (short) 0xffff;
}

